x-pack/plugins/ml/server/lib/alerts/jobs_health_service.ts

```diff
@@ -154,6 +175,62 @@ export function jobsHealthServiceProvider(
           };
         });
     },
+    /**
+     * Returns annotations about delayed data.
+     *
+     * @param jobs
+     * @param timeInterval - Custom time interval provided by the user.
+     * @param docsCount - The threshold for a number of missing documents to alert upon.
+     */
+    async getDelayedDataReport(
+      jobs: MlJob[],
+      timeInterval: string | null,
+      docsCount: number | null
+    ): Promise<DelayedDataResponse[]> {
+      const jobIds = getJobIds(jobs);
+      const datafeeds = await getDatafeeds(jobIds);
+
+      const defaultLookbackInterval = resolveLookbackInterval(jobs, datafeeds!);
+      const earliestMs = getDelayedDataLookbackTimestamp(timeInterval, defaultLookbackInterval);
+
+      let annotations: DelayedDataResponse[] = (
+        await annotationService.getDelayedDataAnnotations({
+          jobIds,
+          earliestMs,
+        })
+      )
+        .map<DelayedDataResponse>((v) => {
+          const match = v.annotation.match(/Datafeed has missed (\d+)\s/);
+          const missedDocsCount = match ? parseInt(match[1], 10) : 0;
+          return {
+            annotation: v.annotation,
+            // end_timestamp is always defined for delayed_data annotation
+            end_timestamp: v.end_timestamp!,
+            missed_docs_count: missedDocsCount,
+            job_id: v.job_id,
+          };
+        })
+        .filter((v) => {
+          // As we retrieved annotations based on the longest bucket span and query delay,
+          // we need to check end_timestamp against appropriate job configuration.
+
+          const job = jobs.find((j) => j.job_id === v.job_id);
+          const datafeed = datafeeds?.find((d) => d.job_id === v.job_id);
+
+          const jobLookbackInterval = resolveLookbackInterval([job!], [datafeed!]);
```

Comment:
👍 this is great.

Get them all in the same query, then filter down appropriately. good stuff.